File: frontend/src/types/dca.ts
export type DCAModelType =
  | "exponential"
  | "hyperbolic"
  | "harmonic"
  | "modified_hyperbolic"
  | "sedm"
  | "duong";

export type FluidType = "oil" | "gas" | "water" | "boe";

export type ChartScale = "linear" | "semi-log" | "log-log";

export interface DCAForecastPoint {
  forecast_date: string;
  time_months: number;
  rate: number;
  cumulative: number;
}

export interface DCAAnalysis {
  id: string;
  well_id: string;
  name: string;
  description: string | null;
  model_type: DCAModelType;
  fluid_type: FluidType;
  start_date: string;
  end_date: string | null;
  parameters: Record<string, number>;
  r_squared: number | null;
  rmse: number | null;
  aic: number | null;
  bic: number | null;
  forecast_months: number;
  economic_limit: number | null;
  eur: number | null;
  remaining_reserves: number | null;
  cum_at_forecast_start: number | null;
  monte_carlo_results: MonteCarloResults | null;
  status: string;
  is_primary: boolean;
  created_at: string;
  updated_at: string;
  forecast_points: DCAForecastPoint[];
}

export interface DCACreate {
  name: string;
  description?: string;
  model_type: DCAModelType;
  fluid_type?: FluidType;
  start_date: string;
  end_date?: string;
  initial_params?: Record<string, number>;
  economic_limit?: number;
  forecast_months?: number;
}

export interface DCAAutoFitRequest {
  well_id: string;
  fluid_type?: FluidType;
  start_date: string;
  end_date?: string;
}

export interface DCAAutoFitResult {
  model_type: DCAModelType;
  parameters: Record<string, number>;
  r_squared: number;
  rmse: number;
  aic: number;
  bic: number;
  eur: number | null;
}

export interface DCAMonteCarloRequest {
  iterations?: number;
  param_distributions: Record<string, ParameterDistribution>;
  economic_limit?: number;
}

export interface ParameterDistribution {
  type: "normal" | "lognormal" | "uniform" | "triangular";
  mean?: number;
  std?: number;
  min?: number;
  max?: number;
  mode?: number;
}

export interface MonteCarloResults {
  p10: number;
  p50: number;
  p90: number;
  mean: number;
  std: number;
  iterations: number;
  eur_distribution?: number[];
}

export const MODEL_LABELS: Record<DCAModelType, string> = {
  exponential: "Exponential",
  hyperbolic: "Hyperbolic",
  harmonic: "Harmonic",
  modified_hyperbolic: "Modified Hyperbolic",
  sedm: "Stretched Exponential (SEDM)",
  duong: "Duong",
};

export const MODEL_PARAMETERS: Record<DCAModelType, string[]> = {
  exponential: ["qi", "di"],
  hyperbolic: ["qi", "di", "b"],
  harmonic: ["qi", "di"],
  modified_hyperbolic: ["qi", "di", "b", "d_min"],
  sedm: ["qi", "tau", "n"],
  duong: ["qi", "a", "m"],
};

export const PARAMETER_UNITS: Record<string, string> = {
  qi: "bbl/d",
  di: "1/month",
  b: "dimensionless",
  d_min: "1/month",
  tau: "months",
  n: "dimensionless",
  a: "dimensionless",
  m: "dimensionless",
};

File: backend/app/schemas/dca.py
from pydantic import BaseModel, Field
from datetime import date, datetime
from uuid import UUID


class DCACreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    description: str | None = None
    model_type: str = Field(
        ...,
        pattern="^(exponential|hyperbolic|harmonic|modified_hyperbolic|sedm|duong)$",
    )
    fluid_type: str = Field(default="oil", pattern="^(oil|gas|water|boe)$")
    start_date: date
    end_date: date | None = None
    initial_params: dict | None = None  # Null = auto-fit
    economic_limit: float = 5.0
    forecast_months: int = Field(default=360, ge=1, le=600)


class DCAUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    parameters: dict | None = None
    economic_limit: float | None = None
    forecast_months: int | None = None


class DCAForecastRequest(BaseModel):
    parameters: dict
    forecast_months: int = 360
    economic_limit: float = 5.0


class DCAMonteCarloRequest(BaseModel):
    iterations: int = Field(default=10000, ge=100, le=100000)
    param_distributions: dict
    economic_limit: float = 5.0


class DCAAutoFitRequest(BaseModel):
    well_id: UUID
    fluid_type: str = "oil"
    start_date: date
    end_date: date | None = None
    forecast_months: int = Field(default=360, ge=1, le=600)
    economic_limit: float = Field(default=5.0, ge=0)


class DCAForecastPointResponse(BaseModel):
    forecast_date: date
    time_months: float
    rate: float
    cumulative: float

    model_config = {"from_attributes": True}


class DCAResponse(BaseModel):
    id: UUID
    well_id: UUID
    name: str
    description: str | None
    model_type: str
    fluid_type: str
    start_date: date
    end_date: date | None
    parameters: dict
    r_squared: float | None
    rmse: float | None
    aic: float | None
    bic: float | None
    forecast_months: int
    economic_limit: float | None
    eur: float | None
    remaining_reserves: float | None
    cum_at_forecast_start: float | None
    monte_carlo_results: dict | None
    status: str
    is_primary: bool
    created_at: datetime
    updated_at: datetime
    forecast_points: list[DCAForecastPointResponse] = []

    model_config = {"from_attributes": True}


class DCAAutoFitResponse(BaseModel):
    model_type: str
    parameters: dict
    r_squared: float
    rmse: float
    aic: float
    bic: float
    eur: float | None
    forecast_points: list[DCAForecastPointResponse] = []
